Updating CallWithRetries to do exponential backoff

moontails opened the pull request:
- setting maximum backoff seconds to be 32 seconds.

moontails committed:
Updating CallWithRetries to do exponential backoff

tensorflow-jenkins commented:
Can one of the admins verify this patch?

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -24,6 +24,8 @@ namespace {
 // In case of failure, every call will be retried kMaxAttempts-1 times.
 constexpr int kMaxAttempts = 4;
+// Maximum backoff time in seconds
+constexpr int maximumBackoffSeconds = 32;
```
Could you please rename this to kMaximumBackoffSeconds in order to be consistent with the naming conventions for constants?

moontails replied:
Did not know that. Will do.

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -39,14 +41,18 @@ bool IsRetriable(Status status) {
 
 Status CallWithRetries(const std::function<Status()>& f) {
   int attempts = 0;
+  int delay;
```
Since the variable 'delay' is not used outside of the 'while' loop, could you please move it inside the while loop?

moontails replied:
I wanted to reuse the delay variable in each loop instead of creating a new one every time.

Should I still go ahead and make the change?

martinwicke replied:
The compiler will take care of optimization if it is beneficial. Please do make the change.

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -39,14 +41,18 @@ bool IsRetriable(Status status) {
     }
     LOG(ERROR) << "The operation resulted in an error and will be retried: "
                << status.ToString();
+    delay = delay > maximumBackoffSeconds ? maximumBackoffSeconds : delay;
```
I wonder if this line and the line 46 could be merged also using std::min, e.g
int delay = std::min(1 << attempts, kMaximumBackoffSeconds) ? This could result in shorter code.

moontails replied:
Good idea!

rinugun reviewed (commented):
Hi! Thanks for improving this.

This is the first time I'm doing a code review on GitHub, so apologies if I'm doing something wrong.

One high level comment: retrying_file_system_test.cc contains test cases for the scenarios when all retry attempts failed. With this change, those tests will take more than 30 seconds to run and might actually time out, depending on what timeout is set for unittests in the test framework (I can imagine it being exactly 30 seconds).

I'd suggest adding an optional constructor parameter initial_delay_seconds, which defaults to 1 but can be overridden to 0 in the unittest, so that the unittest can complete faster.

This code would that use 
initial_delay_seconds << attempts
instead of hardcoding '1' as a constant for the initial delay.

moontails committed:
Addressing CR comments - renaming constant based on convention, moving delay to inside loop and using std::min, adding optional constructor parameter initial_delay_seconds

moontails commented:
@rinugun Thank you for reviewing my PR.

Parameterizing the intial_delay_seconds seems like a good approach, I have addressed the feedback and pushed the changes. Please take a look.

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -24,6 +24,8 @@ namespace {
 
 // In case of failure, every call will be retried kMaxAttempts-1 times.
 constexpr int kMaxAttempts = 4;
+// Maximum backoff time in seconds
```
A minor comment: please add a period at the end for consistency.

moontails replied:
Done

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -40,20 +42,22 @@ bool IsRetriable(Status status) {
 Status CallWithRetries(const std::function<Status()>& f) {
   int attempts = 0;
   while (true) {
+    int delay = std::min(initial_delay_seconds << attempts, kMaximumBackoffSeconds);
```
Minor comments: 
1) could you please add a 'const' modifier?
It's recommended by the style guide, see https://google.github.io/styleguide/cppguide.html#Use_of_const for details. 

2) Could you also move this line to right before line 53, since 'delay' is not used anywhere besides the last line in the loop. This should make the code more readable.

moontails replied:
Regarding 
1. I will add the `constexpr` for delay.

2. I originally added this line here as attempts is incremented at the beginning of the loop. Adding near line 53 would have a slightly uglier way of doing this -
`int delay = std::min(initial_delay_seconds << attempts - 1, kMaximumBackoffSeconds);`

Do you propose to make this change? Or should we increment at the end of the loop and initialize `kMaxAttempts` to 3?
```
constexpr int kMaxAttempts = 3;
...

Status CallWithRetries(const std::function<Status()>& f) {
  int attempts = 0;
  while (true) {
    auto status = f();
    if (!IsRetriable(status) || attempts >= kMaxAttempts) {
      return status;
    }
    LOG(ERROR) << "The operation resulted in an error and will be retried: "
               << status.ToString();
    constexpr int delay = std::min(initial_delay_seconds << attempts, kMaximumBackoffSeconds);
    attempts++;
    sleep(delay);
  }
}
```

rinugun replied:
1. Will 'constexpr' compile though? I thought 'constexpr' has to be computable at compile time. 
2. I think it's fine to move 'attempts++' to make it the last line in the loop (i.e it could be after sleep())
3. In that case its semantics changes from "the number of made attempts" to "the current attempt number", so I suggest renaming 'attempts' to 'attempt' to reflect that
4. Also, 'attempts >= kMaxAttempts" in the 'if' line must be replaced with "attempt > kMaxAttempt" to keep the total amount of allowed attempts the same.

moontails replied:
1. That is correct. Will use`const`

2, 3, 4. Agreed.

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -40,20 +42,22 @@ bool IsRetriable(Status status) {
  public:
-  RetryingRandomAccessFile(std::unique_ptr<RandomAccessFile> base_file)
-      : base_file_(std::move(base_file)) {}
+  RetryingRandomAccessFile(std::unique_ptr<RandomAccessFile> base_file, int delay_seconds = 1)
```
Minor: could you please split the line to keep the line length under 80 chars?
See https://google.github.io/styleguide/cppguide.html#Line_Length for details

moontails replied:
Done

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -86,6 +91,7 @@ class RetryingWritableFile : public WritableFile {
 
  private:
   std::unique_ptr<WritableFile> base_file_;
+  int initial_delay_seconds;
```
Please add a trailing underscore to the name, see this section for details
https://google.github.io/styleguide/cppguide.html#Variable_Names

Please also add a 'const' modifier, as this member is not supposed to be changed in the lifetime of the object.

Also applies to other places where you're adding the new member.

moontails replied:
Will do

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -40,20 +42,22 @@ bool IsRetriable(Status status) {
 Status CallWithRetries(const std::function<Status()>& f) {
```
Did you want to add 'initial_delay_seconds' as a parameter to pass it to CallWithRetries? Otherwise it's not clear to me how initial_delay_seconds inside CallWithRetries is available, or am I missing something?

moontails replied:
Should `initial_delay_seconds` be made a optional arg to `CallWithRetries` with a default of 1? (Rather than it being a member variable and optional constructor parameter)

rinugun replied:
I don't think CallWithRetries is called from the unittest directly, so how would you make sure that initial_delay_seconds is set to 0 for tests?

moontails replied:
makes sense to keep it as optional constructor parameter

moontails replied:
Added as a parameter to pass it to CallWithRetries

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -96,7 +102,7 @@ Status RetryingFileSystem::NewRandomAccessFile(
                                                base_file_system_.get(),
                                                filename, &base_file)));
-  result->reset(new RetryingRandomAccessFile(std::move(base_file)));
+  result->reset(new RetryingRandomAccessFile(std::move(base_file), 0));
```
It looks like if you pass zero for the initial_delay, it will effectively disable exponential backoff. Did you want to pass initial_delay_seconds_ which is a member of RetryingFileSystem?

moontails replied:
My understanding was that you wanted to pass zero for the initial_delay so that the unit tests could run faster?

rinugun replied:
Yes, but zeros should only be passed from the unittest code (as you have implemented in the _test.cc file), whereas the non-test code should still use the actual non-zero values to  have the backoff logic working.

moontails replied:
I see what you mean. Should have updated only in tests. Accidentally updated here as well. My bad. Will revert these changes.

martinwicke commented:
assigning @rinugun to reflect reality. Thanks for reviewing!

moontails committed:
Addressing CR comments - Adding trailing underscore, passing initial_delay_seconds to CallWithRetries, formatting line to be less than 80 chars, updated while loop in CallWithRetries

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -23,7 +23,9 @@ namespace tensorflow {
 
 // In case of failure, every call will be retried kMaxAttempts-1 times.
-constexpr int kMaxAttempts = 4;
+constexpr int kMaxAttempts = 3;
```
Is the change from 4 to 3 intentional?

moontails replied:
`attempt++` happens after the check now. Hence this has been changed from 4 to 3. I need to update the comment however has now it retries exactly `kMaxAttempts` times (if the comparison check is changed to a `>=` or a `==`)

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +39,70 @@ bool IsRetriable(Status status) {
-    attempts++;
     auto status = f();
-    if (!IsRetriable(status) || attempts >= kMaxAttempts) {
+    if (!IsRetriable(status) || attempt > kMaxAttempts) {
```
Sorry, I think I gave you a wrong suggestion. Actually the previous >= condition was more correct here.
So if the attempt number is 1, this means we're making the second attempt. So if kMaxAttempts was 1 and attempt == kMaxAttempt, this means we should fail.

moontails replied:
Is there a preference between `>=` or `>=` ?

rinugun replied:
I think >= feels safer, even though in practice there should be no difference. So I'd prefer >=

moontails replied:
Done

rinugun reviewed (commented):
Looks great, thanks for making all the fixes -- just one minor remaining thing re the kMaxAttempt constant and the way attempts are counted -- sorry for giving you a misleading comment in the past.

moontails commented:
@rinugun Thank you for reviewing my code quickly each time. Learned a few things along the way while helping enhance this :)

moontails committed:
Addressing CR comments - Using the correct comparison check '>='

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -22,8 +22,10 @@ namespace tensorflow {
-// In case of failure, every call will be retried kMaxAttempts-1 times.
-constexpr int kMaxAttempts = 4;
+// In case of failure, every call will be retried kMaxAttempts times.
+constexpr int kMaxAttempts = 3;
```
Ok, I think I got it. It looks like kMaxAttempts has become a "retry count" instead of an "attempt count" that it used to be. So this needs to be fixed to be made consistent again. 

One way would be to rename it to kMaxRetries and rename 'attempt' to 'retry' respectively, keeping the constant value '3', the comment and the rest of the code as it is now.

Another, maybe a simpler way, would be to replace "attempt >= kMaxAttempts" with "attempt >= kMaxAttempts -1" to solve the "off by one" error.  And then the lines 25-26 can be reverted back to the previous state.

What do you think?

moontails replied:
I see how this can be an issue due to the how the variable is named. 

Personally I prefer the first approach. Although it involves more code change, it is easier to read as retry count makes everything slightly more readable.

rinugun replied:
Sounds good to me, thanks!

rinugun replied:
Actually then 'attempt' should become 'retries' (not 'retry') and will represent the amount of retries made so far, so it will make sense that it is  0 in the beginning.

moontails replied:
Agreed. `retries` is just more readable

moontails committed:
Addressing CR comments - renaming to kMaxRetries to be consistent

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -41,18 +41,18 @@ bool IsRetriable(Status status) {
 Status CallWithRetries(const std::function<Status()>& f,
                        const initial_delay_seconds) {
-  int attempt = 0;
+  int retry = 0;
```
Just one last comment -- suggest renaming to 'retries', since it's going to be a retry count now.

moontails replied:
Done

moontails committed:
Addressing CR comments - renaming retry to retries

rinugun reviewed (approved):
Great, thanks for bearing with me!

Let me see if I can figure out how to make your changes get submitted to the repository, I've never done this before.

moontails commented:
@rinugun the pleasure was mine! Your feedback helped polish the code I had submitted. Really appreciate your quick feedback on each commit :) Thank you!

rinugun commented:
Actually I'm not authorized to merge pull requests, so hopefully something good will happen with this PR automatically now when I approved. If not, please ping this thread again, I'll try to investigate.

moontails commented:
I will keep that in mind. Thank you once again @rinugun :)

martinwicke commented:
Jenkins, test this please (@rinugun, you can trigger tests as well)

moontails committed:
Addressing test failures - adding missing closing bracket, adding missing type, inclduing unistd.h for sleep

moontails commented:
Jenkins test this please

martinwicke commented:
Jenkins, test this please

moontails commented:
Error details -
```
tensorflow/core/platform/cloud/retrying_file_system.cc: In member function 'virtual tensorflow::Status tensorflow::RetryingFileSystem::IsDirectory(const string&)':
tensorflow/core/platform/cloud/retrying_file_system.cc:200:76: error: too few arguments to function 'tensorflow::Status tensorflow::{anonymous}::CallWithRetries(const std::function<tensorflow::Status()>&, int)'
       std::bind(&FileSystem::IsDirectory, base_file_system_.get(), dirname));
                                                                            ^
tensorflow/core/platform/cloud/retrying_file_system.cc:42:8: note: declared here
 Status CallWithRetries(const std::function<Status()>& f,
        ^
tensorflow/core/platform/cloud/retrying_file_system.cc:201:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
```

I checked and see that the `IsDirectory` method is not overridden. But I fail to see why this causes an error as the base class implementation does not wrap any call with `CallWithRetries`.

martinwicke commented:
Not quite yet. :)

rinugun commented:
@moontails, I think your local version of the repository is outdated, the function is overridden in the latest version on GitHub https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/cloud/retrying_file_system.cc#L170

moontails commented:
@rinugun I will sync with master and make the changes.

moontails committed:
Merge remote-tracking branch 'upstream/master'

moontails committed:
Syncing with master and updating IsDirectory method to pass the initial_delay_seconds_

rinugun commented:
Jenkins, test this please

moontails commented:
@rinugun @martinwicke the tests werent trigerred

martinwicke commented:
Jenkins, test this please?

moontails commented:
looks like the failed jobs lost connection?

jhseu commented:
Jenkins, test this please

jhseu commented:
Hmm, that failure looks unrelated. Can you try rebasing on top of the latest master and we'll retest?

moontails commented:
Sure will do. Do you want me to rebase or merge with latest master?

Best Regards,
Balachander Ramachandran

On Sep 27, 2016 1:13 PM, "Jonathan Hseu" <notifications@github.com> wrote:

> Hmm, that failure looks unrelated. Can you try rebasing on top of the
> latest master and we'll retest?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/4501#issuecomment-249983540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEL334CZH4uxTzvd4rNapE2-TualfS6sks5quXjggaJpZM4KCXcW>
> .
>

jhseu commented:
Merge is fine too (and probably easier). We have some changes in master that are needed for tests to work.

rinugun commented:
@moontails, I just realized that sleep() might not be the canonical way of making a delay, e.g it might not be supported on Windows. I was wondering if you could replace it with 
Env::Default()->SleepForMilliseconds, e.g see this example:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/cloud/gcs_file_system.cc#L379

This also prepares the code for the subsequent implementation of random millisecond part, that we discussed in another thread with @davidj-github.

Also I had another comment related to this, even though this contradicts to what I suggested before. Instead of making 'initial_delay' a ctor parameter, you could have 'Env* env' a ctor parameter instead, which would default to Env::Default(). This way you could provide a fake implementation of SleepForMilliseconds from the test code. The fake implementation would return immediately, but would record how it was called.
The benefit of the latter is that the test code can still work fast, but also check the sleep intervals that were passed to SleepForMilliseconds from the main code and make sure they all look as expected.

moontails committed:
Merge remote-tracking branch 'upstream/master'

moontails commented:
@jhseu I have merged the latest changes from master. Could you please trigger the tests?

jhseu commented:
Sure, but mind making the requested changes before we commit?
Jenkins, test this please

moontails committed:
Sleeping for additional random milliseconds and Updating GetChildrenRecursively method to pass the initial_delay_seconds_

moontails committed:
Merging and Resolving conflicts

moontails commented:
@jhseu @rinugun committed the changes. Please review and trigger tests!

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -14,16 +14,21 @@ limitations under the License.
-constexpr int kMaxAttempts = 4;
+// In case of failure, every call will be retried kMaxRetries times.
+constexpr int kMaxRetries = 3;
+// Maximum backoff time in seconds.
```
Please update the comment.

moontails replied:
Done

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -14,16 +14,21 @@ limitations under the License.
+// In case of failure, every call will be retried kMaxRetries times.
+constexpr int kMaxRetries = 3;
+// Maximum backoff time in seconds.
+constexpr int kMaximumBackoffMilliSeconds = 32000;
```
A minor thing, but "millisecond" is one word, so it should be kMaximumBackoffMilliseconds (small "s")

rinugun replied:
Also, please update to int64 to be on the safe side.

moontails replied:
Done

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +42,81 @@ bool IsRetriable(Status status) {
+
+  const int delay_milliseconds = delay_seconds * 1000;
+  const int random_milliseconds = dist(gen);
+  const int delay = std::min(delay_milliseconds + random_milliseconds,
```
Another minor comment: since 'delay' here in fact also means "delay in milliseconds", this now conflicts with the variable at line 50. To resolve this, I suggest renaming 'delay' to 'delay_milliseconds' and inlining the previous 'delay_milliseconds', e.g

const int64 delay = std::min(delay_seconds * 1000 + random_milliseconds)

rinugun replied:
Sorry, what I meant was
const int64 delay_milliseconds = std::min(delay_seconds * 1000 + random_milliseconds,
                                                                      kMaximumBackoffMilliseconds);

moontails replied:
Yup got it, Done

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +42,81 @@ bool IsRetriable(Status status) {
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<> dist(0, 999);
+
+  const int delay_milliseconds = delay_seconds * 1000;
```
Could you update the types to  int64 to be sure, also the next line.

moontails replied:
Done

rinugun commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +42,81 @@ bool IsRetriable(Status status) {
+    if (!IsRetriable(status) || retries >= kMaxRetries) {
       return status;
     }
     LOG(ERROR) << "The operation resulted in an error and will be retried: "
```
I wonder if you could also print the delay in seconds in the LOG message, e.g 
const int delay_seconds = initial_delay_seconds << retries;
... will be retried after " << delay_seconds << " sec:"

moontails replied:
Done

moontails committed:
Adressing CR comments - updating comment

moontails committed:
Adressing CR comments - using int64, updating log message

rinugun commented:
Sorry for not sending all comments at once -- hit a wrong button on Github.
Please see my other comments as well.

On Tue, Sep 27, 2016 at 9:32 PM, Balachander Ramachandran <
notifications@github.com> wrote:

> *@moontails* commented on this pull request.
> ------------------------------
>
> In tensorflow/core/platform/cloud/retrying_file_system.cc
> <https://github.com/tensorflow/tensorflow/pull/4501>:
>
> >  #include "tensorflow/core/lib/core/errors.h"
>  #include "tensorflow/core/platform/file_system.h"
>
>  namespace tensorflow {
>
>  namespace {
>
> -// In case of failure, every call will be retried kMaxAttempts-1 times.
> -constexpr int kMaxAttempts = 4;
> +// In case of failure, every call will be retried kMaxRetries times.
> +constexpr int kMaxRetries = 3;
> +// Maximum backoff time in seconds.
>
> Done
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/4501>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABnZsn9f8SkYsVLLGybXBuw5Iywy8QYeks5que3ngaJpZM4KCXcW>
> .
>

moontails commented:
@rinugun I have addressed all your comments

rinugun reviewed (approved):
Thanks, looks great!

moontails committed:
Merge remote-tracking branch 'upstream/master'

moontails commented:
@jhseu @martinwicke could one of you please trigger the tests

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +42,81 @@ bool IsRetriable(Status status) {
+  const int64 delay_ms = std::min(delay_seconds * 1000 + random_ms,
+                                  kMaximumBackoffMilliseconds);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(delay_ms));
```
Please use Env::SleepForMicroseconds instead:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/env.h#L220

There are internal reasons why we need to use a non-standard sleep call.

moontails replied:
Done

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +42,81 @@ bool IsRetriable(Status status) {
-  int attempts = 0;
+void WaitBeforeRetry(const int delay_seconds) {
+  std::random_device rd;
+  std::mt19937 gen(rd());
```
Please call Rand64() instead:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/random/random.h

It uses a static device, so won't reconstruct every call. Something like `const int64 random_ms = Rand64() % 1000`

moontails replied:
Done

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +42,81 @@ bool IsRetriable(Status status) {
+}
+
+Status CallWithRetries(const std::function<Status()>& f,
+                       const int initial_delay_seconds) {
```
Can you pass around microseconds for consistency with the rest of tensorflow?

moontails replied:
Done

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -133,49 +168,58 @@ bool RetryingFileSystem::FileExists(const string& fname) {
   return CallWithRetries(std::bind(&FileSystem::GetChildren,
-                                   base_file_system_.get(), dir, result));
+                                   base_file_system_.get(), dir, result),
+                         initial_delay_seconds_);
```
nit: here and below, please align

moontails replied:
Isn't the alignment right here? `std::bind(&FileSystem::GetChildren, base_file_system_.get(), dir, result)` is one parameter and `initial_delay_seconds_` is another.

Parameters to `std::bind` are aligned and parameters to `CallWithRetries` are algined

jhseu replied:
Oh, you're right, ignore this comment.

jhseu reviewed (commented):
Sorry, I'm a little pickier than @rinugun :)

A few more comments.

moontails committed:
Adressing CR comments - using SleepForMicroseconds, changed to microseconds and New64

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -14,16 +14,23 @@ limitations under the License.
 ==============================================================================*/
 
 #include "tensorflow/core/platform/cloud/retrying_file_system.h"
+#include <chrono>
```
nit: remove chrono random and thread includes?

moontails replied:
my bad, should have cleaned up unused imports

moontails committed:
Adressing CR comments - Removing unused imports

moontails committed:
Merge remote-tracking branch 'upstream/master'

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.h:
```diff
@@ -16,6 +16,8 @@ limitations under the License.
 #ifndef TENSORFLOW_CORE_PLATFORM_RETRYING_FILE_SYSTEM_H_
 #define TENSORFLOW_CORE_PLATFORM_RETRYING_FILE_SYSTEM_H_
 
+#include <unistd.h>
```
Delete?

moontails replied:
Done

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -16,14 +16,18 @@ limitations under the License.
-constexpr int kMaxAttempts = 4;
+// In case of failure, every call will be retried kMaxRetries times.
+constexpr int kMaxRetries = 3;
+// Maximum backoff time in milliseconds.
```
milliseconds -> microseconds

moontails replied:
Done

jhseu commented on tensorflow/core/platform/cloud/retrying_file_system.cc:
```diff
@@ -37,55 +41,76 @@ bool IsRetriable(Status status) {
-Status CallWithRetries(const std::function<Status()>& f) {
-  int attempts = 0;
+void WaitBeforeRetry(const int64 delay_micros) {
+  const int64 random_micros = random::New64() % 1000000L;
```
Here and below: L is not necessary for these constants (and isn't sufficient to make it an int64)

moontails replied:
so i should remove L ?

jhseu replied:
Yes, please

moontails replied:
quick question should i be removing it from every place i have used? (like `kMaximumBackoffMicroseconds` , constructor args default value etc... )

moontails replied:
Removed it everywhere I introduced it

jhseu reviewed (commented):
Looks good. A few more nitpicks :)

moontails committed:
Adressing CR comments - removing L suffix for int64 and removing unused imports

jhseu commented:
Jenkins, test this please

moontails committed:
Updating delay to 0 in new tests from upstream/master to avoid timeouts

jhseu commented:
Jenkins, test this please

moontails commented:
@jhseu @rinugun thank you for all the feedback during this code review and responding to my questions so quickly each time :)

jhseu commented:
Thanks you for your contribution!
